### ISSUE TITLE
__set__ value type ignored when assigning to self.attribute

### ISSUE THREAD

**Issue by sbidoul** [microsoft/pyright#4024]
**Describe the bug**
The value type of a descriptor `__set__` method is ignored in some circumstances?
Not sure how to describe this in words. Hopefully the reproducer below makes sense.

**To Reproduce**

Assume the following:

```python
class IntField:
    def __set__(self, instance: object, value: int) -> None:
        ...

class A:

    field = IntField()

    def f(self) -> None:
        self.field = "a"  # (1)

A().field = "a"  # (2)
```

At line (1) no error is reported.
At line (2), pyright correctly reports this:

```
Cannot assign member "field" for type "A"
"Literal['a']" is incompatible with "int" (reportGeneralTypeIssues)
```

**Expected behavior**
A similar error should be reported for (1).

**Comment by erictraut** [comment]
The problem is that you haven't provided a type declaration for `field`. That means its type is inferred. Within the class implementation, an assignment of any value is allowed because there is no declared type. The inferred type of the symbol is the union of all assigned types. In your case, the inferred type of `A.field` is `IntField | str`.

Consider the following example, which doesn't involve a descriptor:

```python
class A:
    x = None
    def f(self, x: int):
        self.x = x

a = A()
reveal_type(a.x)  # int | None
a.x = 3.4 # float is incompatible with int | None
```

To make this work as you've intended, you need to provide a type declaration like this:

```python
field: IntField = IntField()
```

**Comment by erictraut** [comment]
For more details about how pyright uses type declarations and type inference, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#type-inference).

**Comment by sbidoul** [comment]
@erictraut thanks for the quick response. I would agree if there was no `__set__` method on IntType.

I simplified my example even more. The inferred type at line (1) is `IntField`, which has a `__set__` method that accepts `int` only, so assigning a `str` to it amount to calling `__set__` with an incorrect type, no? So I would think it should flag that instead of also inferring `str` as a possible type for `field`.

FWIW, here is what mypy says, which seems correct to me.

```
class IntField:
    def __set__(self, instance: object, value: int) -> None:
        ...

class A:

    field = IntField()

    def f(self) -> None:
        reveal_type(self.field)
        self.field = "a"  # (1)
```

```
test.py:10: note: Revealed type is "test.IntField"
test.py:11: error: Incompatible types in assignment (expression has type "str", variable has type "int")
Found 1 error in 1 file (checked 1 source file)
```

**Comment by erictraut** [comment]
Mypy treats assignments as implied type declarations in many cases (although it's somewhat inconsistent and there are cases where it does not). This behavior causes many false positives, so pyright's approach is different.

For pyright, you must provide a type annotation if you want to declare the type of a symbol. If you don't provide a type declaration, pyright uses consistent rules to infer the type of the symbol as outlined in the documentation. In particular, types are inferred based on all assignments within the class implementation. Pyright is working as designed here. This behavior is intended, and there is no bug to fix.

In a simple example like the one you've provided, it might seem like it's obvious that the type of `field` should be inferred as `IntField` rather than `IntField | str`, but consider a slightly more complex case.

```python
class A1:
    if some_condition:
        field = IntField()
    else:
        field = 0
```

or 

```python
class A2:
    @classmethod
    def func1(cls):
        cls.field = IntField()
    
    @classmethod
    def func2(cls):
        cls.field = 0
```

As I mentioned above, the recommended approach is to provide a type declaration for the class variable `field`: `field: IntField = IntField()`.

**Comment by sbidoul** [comment]
Ok, I understand the type inference mechanism, but allow me a last question.

In my original example, the error is `"Literal['a']" is incompatible with "int" (reportGeneralTypeIssues)`.
Notice it says **int**, while nowhere there is a `int` assigned to `field`.
So there, it seems pyright did look at the type of the `value` argument of `__set__()` to infer that assigning to `field` goes via the descriptor and must accept only a `int`.

So it seems that there is some special case treatment of assignment in presence of `__set__` descriptors?

In other words, why do I get the expected error at one assignment and not the other?

Note I understand the proposed workaround. I'm just digging a little bit to find out if I can avoid the redundancy.

**Comment by erictraut** [comment]
This error is related to an unspecified aspect of the Python type system — namely, how to enforce assigned values when the type of the assignment target is a union that contains a descriptor and a non-descriptor subtype. This is a very unusual situation — one that I've never come across before, so it's unsurprising that this is an unspecified condition.

Consider the following code. Note that pyright and mypy handle these cases differently. Since this is an unspecified part of the type system, I think either behavior could be considered "correct". Likewise, I could argue that both are "incorrect", and no errors should be generated for this code.

```python
class IntField:
    def __set__(self, instance: object, value: int) -> None:
        ...

class StrField:
    def __set__(self, instance: object, value: str) -> None:
        ...

class A:
    f1: int | str
    f2: int | StrField
    f3: IntField | str
    f4: IntField | StrField

a = A()
a.f1 = ""  # pyright: no error, mypy: no error
a.f2 = ""  # pyright: no error, mypy: error
a.f3 = ""  # pyright: error,    mypy: no error
a.f4 = ""  # pyright: error,    mypy: error
```

Since this is an unspecified part of the spec and it's such an odd edge case, I don't think it's worth making a change to either pyright or mypy.

**Comment by sbidoul** [comment]
> an odd edge case

I'm trying to type-enable an existing ORM without being too disruptive, nor introducing an overly verbose notation.

> a union that contains a descriptor and a non-descriptor subtype

I'm wondering if it would make sense for `self.field = "a"` to consider that `cls.field` is a descriptor with `__set__` and from that, avoid inferring a Union type in the first place.

**Comment by erictraut** [comment]
The normal type inference rule used by pyright is summarized as follows:

1. For each value assigned to the symbol, evaluate the type of the assigned value
2. Combine the types evaluated in step 1 into a union

This is very straightforward, easy to explain, and easy to understand. It also composes well with other functionality and behaviors.

What you're suggesting is that we change the above algorithm to the following:

1. For each value assigned to the symbol, evaluate the type of the assigned expression
2. If the symbol is a variable and is scoped to a class:
  a. Determine which subset of the assignments from step 1 are assigned through a class reference (or directly in the class body) versus an object reference, and combine the former into a union
  b. If any of the subtypes within the union from step 2a are writable descriptor objects (i.e. objects that support a `__get__` and `__set__` method), use the union from step 2a as the inferred type
3. Combine the types evaluated in step 1 into a union

Special cases like this tend to have cascade effects in a type checker, and they require yet more special casing as they break assumptions in other parts of the code. For example, to handle the above special-case change in the inference logic, there would need to be other changes to pyright's logic for enforcing writes to unannotated instance variables from within a class implementation.

I'm not convinced that this added level of complexity and special-case logic is a good idea in this case.

For now, I'm going to leave the current behavior in place and leave this issue closed. We may reconsider this in the future based on signal from other pyright users.

**Comment by sbidoul** [comment]
Thanks a lot for your time and detailed analysis. I understand this case would add complexity to the type checker.

I'm cross-referencing https://github.com/microsoft/pyright/issues/3628#issuecomment-1169089086 which has a discussion of descriptors in the context of SQLAlchemy. The ORM I'm playing with a different one (Odoo), but the problem space is the same.

Since mypy and pyre seem to support this it would be cool if pyright would do too, since it enables a very interesting pattern and concise notation for declaring typed ORM fields.
